ftl/python/layer_builder.py

```diff
@@ -214,6 +214,93 @@ def _log_cache_result(self, hit):
                 key=self.GetCacheKey()))
 
 
+class PipfileLayerBuilder(RequirementsLayerBuilder):
+    def __init__(self,
+                 ctx=None,
+                 descriptor_files=None,
+                 pkg_descriptor=None,
+                 pkg_dir=None,
+                 dep_img_lyr=None,
+                 wheel_dir=None,
+                 venv_dir=None,
+                 pip_cmd=None,
+                 venv_cmd=None,
+                 cache=None):
+        super(RequirementsLayerBuilder, self).__init__()
+        self._ctx = ctx
+        self._pkg_dir = pkg_dir
+        self._wheel_dir = wheel_dir
+        self._venv_dir = venv_dir
+        self._pip_cmd = pip_cmd
+        self._venv_cmd = venv_cmd
+        self._descriptor_files = descriptor_files
+        self._dep_img_lyr = dep_img_lyr
+        self._cache = cache
+        self._pkg_descriptor = pkg_descriptor
+
+    def GetCacheKeyRaw(self):
+        return "%s %s %s" % (self._pkg_descriptor[0], self._pkg_descriptor[1],
+                             self._dep_img_lyr.GetCacheKeyRaw())
+
+    def _log_cache_result(self, hit):
+        if hit:
+            cache_str = constants.PHASE_2_CACHE_HIT
+        else:
+            cache_str = constants.PHASE_2_CACHE_MISS
+        logging.info(
+            cache_str.format(
+                key_version=constants.CACHE_KEY_VERSION,
+                language='PYTHON',
+                package_name=self._pkg_descriptor[0],
+                package_version=self._pkg_descriptor[1],
+                key=self.GetCacheKey()))
+
+    def BuildLayer(self):
+        cached_img = None
+        if self._cache:
+            with ftl_util.Timing('Checking cached pkg layer'):
+                key = self.GetCacheKey()
+                cached_img = self._cache.Get(key)
+                self._log_cache_result(False if cached_img is None else True)
+        if cached_img:
+            self.SetImage(cached_img)
+        else:
+            with ftl_util.Timing('Installing pip packages'):
+                self._pip_install(' '.join(self._pkg_descriptor))
+
+            with ftl_util.Timing('Resolving whl paths'):
+                whls = self._resolve_whls()
+                pkg_dirs = [self._whl_to_fslayer(whl) for whl in whls]
+                if len(pkg_dirs) != 1:
```

Comment:
this seems like a strange way to do this check. maybe check `len(whls)` before calling `_whl_to_fslayer()` and then skip creating the list?

Reply:
done